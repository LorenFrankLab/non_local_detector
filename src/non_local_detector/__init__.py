from non_local_detector.models import (
    ContFragClusterlessClassifier,  # noqa
    ContFragSortedSpikesClassifier,  # noqa
    MultiEnvironmentSortedSpikesClassifier,  # noqa
    MultiEnvironmentClusterlessClassifier,  # noqa
    NonLocalClusterlessDetector,  # noqa
    NonLocalSortedSpikesDetector,  # noqa
    ClusterlessDecoder,  # noqa
    SortedSpikesDecoder,  # noqa
)

try:
    from ._version import __version__
except ImportError:
    pass
