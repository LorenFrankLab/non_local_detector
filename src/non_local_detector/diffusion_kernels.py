from typing import Any

import jax
import jax.numpy as jnp
import jax.scipy.linalg
import networkx as nx
import numpy as np


def _assign_gaussian_weights_from_distance(G: nx.Graph, bandwidth_sigma: float) -> None:
    """
    Overwrites each edge's "weight" attribute with
        w_uv = exp( - (distance_uv)^2 / (2 * sigma^2) ).
    Assumes each edge already has "distance" = Euclidean length.
    """
    two_sigma2 = 2.0 * (bandwidth_sigma**2)
    for u, v, data in G.edges(data=True):
        d = data.get("distance", None)
        if d is None:
            raise KeyError(f"Edge ({u},{v}) has no 'distance' attribute.")
        # compute the RBF weight
        data["weight"] = float(np.exp(-(d * d) / two_sigma2))


def compute_diffusion_kernels(
    graph: nx.Graph,
    bandwidth_sigma: float,
) -> jnp.ndarray:
    """
    Computes diffusion kernels for all valid bins using the
    Graph Laplacian method and matrix exponential.

    Parameters
    ----------
    graph : nx.Graph
        The graph representing the track, where nodes are bins and edges are connections.
    bandwidth_sigma : float
        The bandwidth of the Gaussian kernel. This controls the spread of the kernel.

    Returns
    -------
    kernel_matrix : jnp.ndarray, shape (n_bins, n_bins)
        The diffusion kernel matrix for the bins.
        Therepresents the amount of concentration (or probability mass, if u represents probability)
        that has flowed from node j to node i after time t.
    """

    n_bins = graph.number_of_nodes()
    _assign_gaussian_weights_from_distance(graph, bandwidth_sigma)
    laplacian_full = jnp.array(
        nx.laplacian_matrix(graph, nodelist=range(n_bins), weight="weight").toarray(),
        dtype=jnp.float32,
    )

    exponent_coefficient = bandwidth_sigma**2 / 2.0
    kernel_matrix = jax.scipy.linalg.expm(-exponent_coefficient * laplacian_full)

    # Clip to non-negative values
    kernel_matrix = jnp.clip(kernel_matrix, a_min=0.0, a_max=None)

    # Normalize the kernel matrix so the sum of each column is 1.
    col_sums = kernel_matrix.sum(axis=0, keepdims=True)
    kernel_matrix = jnp.where(col_sums == 0.0, 0.0, kernel_matrix / col_sums)

    return kernel_matrix
