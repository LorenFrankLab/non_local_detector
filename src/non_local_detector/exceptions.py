"""Custom exceptions for non_local_detector.

This module provides specialized exceptions that give helpful, actionable error
messages to guide users toward correct usage. Following Raymond Hettinger's
principle: "Error messages should be your best documentation."

Usage Guidelines
----------------
Choose the appropriate exception type based on when and why the error occurs:

- **ValidationError**: Input parameters don't meet requirements (raised during
  initialization or before model fitting). Use when user-provided values are
  invalid, have wrong shapes, or violate constraints.

- **FittingError**: Problems occur during model training/fitting. Use when the
  fitting process encounters errors, but the inputs were valid.

- **ConvergenceError**: Iterative algorithms fail to converge (specialized
  FittingError). Use when EM algorithm, gradient descent, or other iterative
  methods don't reach convergence criteria.

- **ConfigurationError**: Incompatible parameter combinations or missing
  required configuration. Use when parameters are individually valid but
  incompatible with each other.

- **DataError**: Data quality issues like NaN, Inf, or wrong structure. Use
  when the data values (not just shapes) are problematic.

All exceptions inherit from **NonLocalDetectorError**, allowing users to catch
any package-specific error with a single except clause.

Examples
--------
>>> from non_local_detector.exceptions import ValidationError, NonLocalDetectorError
>>>
>>> # Raising a validation error with structured message:
>>> try:
...     raise ValidationError(
...         "Array shape mismatch",
...         expected="shape (100, 2)",
...         got="shape (100, 3)",
...         hint="Position array should have exactly 2 columns (x, y)",
...         example="position = np.array([[1, 2], [3, 4]])"
...     )
... except ValidationError as e:
...     print(e)
...
>>> # Catching all package errors:
>>> try:
...     # Some operation that might fail
...     pass
... except NonLocalDetectorError as e:
...     # Handle any package error
...     print(f"Package error: {e}")
"""


class NonLocalDetectorError(Exception):
    """Base exception for all non_local_detector errors.

    All custom exceptions in this package inherit from this base class,
    making it easy to catch any package-specific error.

    Examples
    --------
    >>> try:
    ...     # Any non_local_detector operation
    ...     pass
    ... except NonLocalDetectorError:
    ...     # Handle any package error
    ...     pass
    """

    pass


class ValidationError(NonLocalDetectorError):
    """Raised when input validation fails.

    This exception indicates that user-provided data doesn't meet the
    required format, shape, or constraints. The error message should
    explain what was expected, what was received, and how to fix it.

    Parameters
    ----------
    message : str
        Description of what went wrong
    expected : str, optional
        What was expected (for structured error messages)
    got : str, optional
        What was actually received
    hint : str, optional
        Actionable suggestion for fixing the error
    example : str, optional
        Code snippet showing correct usage

    Examples
    --------
    >>> raise ValidationError(
    ...     "Array dimensions don't match",
    ...     expected="shape (100, 2)",
    ...     got="shape (100, 3)",
    ...     hint="Check that position array has exactly 2 columns (x, y)"
    ... )
    """

    def __init__(
        self,
        message: str,
        expected: str | None = None,
        got: str | None = None,
        hint: str | None = None,
        example: str | None = None,
    ):
        """Initialize ValidationError with structured message components."""
        parts = [message]

        if expected is not None:
            parts.append(f"\nExpected: {expected}")

        if got is not None:
            parts.append(f"Got: {got}")

        if hint is not None:
            parts.append(f"\nHint: {hint}")

        if example is not None:
            parts.append(f"\nExample:\n{example}")

        super().__init__("\n".join(parts))


class FittingError(NonLocalDetectorError):
    """Raised when model fitting fails.

    This exception indicates that the model fitting process encountered
    an error, such as convergence failure, invalid training data, or
    numerical instability.

    Parameters
    ----------
    message : str
        Description of what went wrong during fitting
    hint : str, optional
        Actionable suggestion for fixing the error

    Examples
    --------
    >>> raise FittingError(
    ...     "Model failed to converge after 1000 iterations",
    ...     hint="Try increasing discrete_transition_concentration or check for NaN values in data"
    ... )
    """

    def __init__(self, message: str, hint: str | None = None) -> None:
        """Initialize FittingError with optional hint.

        Parameters
        ----------
        message : str
            Description of what went wrong during fitting
        hint : str, optional
            Actionable suggestion for fixing the error
        """
        if hint is not None:
            message = f"{message}\n\nHint: {hint}"
        super().__init__(message)


class ConfigurationError(NonLocalDetectorError):
    """Raised when configuration is invalid or inconsistent.

    This exception indicates that model parameters are incompatible
    or that required configuration is missing.

    Parameters
    ----------
    message : str
        Description of the configuration problem
    hint : str, optional
        Actionable suggestion for fixing the configuration

    Examples
    --------
    >>> raise ConfigurationError(
    ...     "Cannot use 'uniform' initial conditions with 'custom' transitions",
    ...     hint="Either use UniformInitialConditions() or provide compatible transitions"
    ... )
    """

    def __init__(self, message: str, hint: str | None = None) -> None:
        """Initialize ConfigurationError with optional hint.

        Parameters
        ----------
        message : str
            Description of the configuration problem
        hint : str, optional
            Actionable suggestion for fixing the configuration
        """
        if hint is not None:
            message = f"{message}\n\nHint: {hint}"
        super().__init__(message)


class ConvergenceError(FittingError):
    """Raised when an iterative algorithm fails to converge.

    This is a specialized fitting error for convergence failures,
    providing specific guidance about iteration limits and tolerances.

    Parameters
    ----------
    message : str
        Description of the convergence failure
    iterations : int, optional
        Number of iterations attempted
    tolerance : float, optional
        Convergence tolerance that wasn't met
    hint : str, optional
        Actionable suggestion for fixing convergence issues

    Examples
    --------
    >>> raise ConvergenceError(
    ...     "EM algorithm did not converge",
    ...     iterations=1000,
    ...     tolerance=1e-4,
    ...     hint="Try increasing max_iterations or relaxing convergence tolerance"
    ... )
    """

    def __init__(
        self,
        message: str,
        iterations: int | None = None,
        tolerance: float | None = None,
        hint: str | None = None,
    ) -> None:
        """Initialize ConvergenceError with iteration details.

        Parameters
        ----------
        message : str
            Description of the convergence failure
        iterations : int, optional
            Number of iterations attempted
        tolerance : float, optional
            Convergence tolerance that wasn't met
        hint : str, optional
            Actionable suggestion for fixing convergence issues
        """
        if iterations is not None:
            message = f"{message} (iterations: {iterations})"
        if tolerance is not None:
            message = f"{message} (tolerance: {tolerance})"

        super().__init__(message, hint=hint)


class DataError(NonLocalDetectorError):
    """Raised when input data has problems (NaN, Inf, wrong structure).

    This exception indicates issues with the actual data values rather
    than structural validation issues.

    Parameters
    ----------
    message : str
        Description of the data problem
    data_name : str, optional
        Name of the problematic data variable
    hint : str, optional
        Actionable suggestion for fixing the data issue

    Examples
    --------
    >>> raise DataError(
    ...     "Found NaN values in spike data",
    ...     data_name="multiunit_firing_rates",
    ...     hint="Check for missing data or use np.nan_to_num() to replace NaN values"
    ... )
    """

    def __init__(
        self, message: str, data_name: str | None = None, hint: str | None = None
    ) -> None:
        """Initialize DataError with optional data name and hint.

        Parameters
        ----------
        message : str
            Description of the data problem
        data_name : str, optional
            Name of the problematic data variable
        hint : str, optional
            Actionable suggestion for fixing the data issue
        """
        if data_name is not None:
            message = f"{message} (data: {data_name})"
        if hint is not None:
            message = f"{message}\n\nHint: {hint}"
        super().__init__(message)
