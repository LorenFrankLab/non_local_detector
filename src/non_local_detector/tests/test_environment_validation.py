"""Tests for Environment validation."""

import networkx as nx
import numpy as np
import pytest

from non_local_detector import Environment
from non_local_detector.exceptions import ConfigurationError, ValidationError


class TestEnvironmentBasicValidation:
    """Test basic parameter validation for Environment."""

    def test_valid_environment_minimal(self):
        """Test that minimal valid environment works."""
        env = Environment()
        assert env.place_bin_size == 2.0
        assert env.bin_count_threshold == 0

    def test_valid_environment_with_bin_size(self):
        """Test environment with custom bin size."""
        env = Environment(place_bin_size=5.0)
        assert env.place_bin_size == 5.0

    def test_valid_environment_with_tuple_bin_size(self):
        """Test environment with multi-dimensional bin size."""
        env = Environment(place_bin_size=(5.0, 10.0))
        assert env.place_bin_size == (5.0, 10.0)

    def test_negative_place_bin_size_raises(self):
        """Test that negative place_bin_size raises."""
        with pytest.raises(ValidationError) as exc_info:
            Environment(place_bin_size=-1.0)
        assert "place_bin_size" in str(exc_info.value)
        assert "positive" in str(exc_info.value).lower()

    def test_zero_place_bin_size_raises(self):
        """Test that zero place_bin_size raises."""
        with pytest.raises(ValidationError) as exc_info:
            Environment(place_bin_size=0.0)
        assert "place_bin_size" in str(exc_info.value)

    def test_negative_tuple_place_bin_size_raises(self):
        """Test that negative values in tuple place_bin_size raise."""
        with pytest.raises(ValidationError) as exc_info:
            Environment(place_bin_size=(5.0, -1.0))
        assert "place_bin_size[1]" in str(exc_info.value)

    def test_negative_bin_count_threshold_raises(self):
        """Test that negative bin_count_threshold raises."""
        with pytest.raises(ValidationError) as exc_info:
            Environment(bin_count_threshold=-1)
        assert "bin_count_threshold" in str(exc_info.value)


class TestTrackGraphValidation:
    """Test validation of track_graph parameter."""

    def test_track_graph_requires_networkx_graph(self):
        """Test that track_graph must be a NetworkX Graph."""
        with pytest.raises(ValidationError) as exc_info:
            Environment(track_graph={"node": "value"})
        assert "networkx" in str(exc_info.value).lower()

    def test_track_graph_requires_nodes(self):
        """Test that track_graph must have nodes."""
        with pytest.raises(ValidationError) as exc_info:
            Environment(track_graph=nx.Graph())
        assert "nodes" in str(exc_info.value).lower()

    def test_track_graph_requires_edges(self):
        """Test that track_graph must have edges."""
        g = nx.Graph()
        g.add_node(0, pos=(0, 0))
        g.add_node(1, pos=(1, 1))
        with pytest.raises(ValidationError) as exc_info:
            Environment(track_graph=g)
        assert "edges" in str(exc_info.value).lower()

    def test_track_graph_nodes_require_pos_attribute(self):
        """Test that all nodes must have 'pos' attribute."""
        g = nx.Graph()
        g.add_node(0, pos=(0, 0))
        g.add_node(1)  # Missing pos
        g.add_edge(0, 1, distance=1.0, edge_id=0)
        with pytest.raises(ValidationError) as exc_info:
            Environment(track_graph=g, edge_order=[(0, 1)], edge_spacing=0.0)
        assert "pos" in str(exc_info.value)
        assert "attribute" in str(exc_info.value).lower()

    def test_track_graph_edges_require_distance_attribute(self):
        """Test that all edges must have 'distance' attribute."""
        g = nx.Graph()
        g.add_node(0, pos=(0, 0))
        g.add_node(1, pos=(1, 1))
        g.add_edge(0, 1, edge_id=0)  # Missing distance
        with pytest.raises(ValidationError) as exc_info:
            Environment(track_graph=g, edge_order=[(0, 1)], edge_spacing=0.0)
        assert "distance" in str(exc_info.value)
        assert "attribute" in str(exc_info.value).lower()

    def test_track_graph_edges_require_edge_id_attribute(self):
        """Test that all edges must have 'edge_id' attribute."""
        g = nx.Graph()
        g.add_node(0, pos=(0, 0))
        g.add_node(1, pos=(1, 1))
        g.add_edge(0, 1, distance=1.0)  # Missing edge_id
        with pytest.raises(ValidationError) as exc_info:
            Environment(track_graph=g, edge_order=[(0, 1)], edge_spacing=0.0)
        assert "edge_id" in str(exc_info.value)
        assert "attribute" in str(exc_info.value).lower()

    def test_track_graph_requires_edge_order(self):
        """Test that track_graph requires edge_order parameter."""
        g = nx.Graph()
        g.add_node(0, pos=(0, 0))
        g.add_node(1, pos=(1, 1))
        g.add_edge(0, 1, distance=1.0, edge_id=0)
        with pytest.raises(ConfigurationError) as exc_info:
            Environment(track_graph=g)
        assert "edge_order" in str(exc_info.value)

    def test_track_graph_requires_edge_spacing(self):
        """Test that track_graph requires edge_spacing parameter."""
        g = nx.Graph()
        g.add_node(0, pos=(0, 0))
        g.add_node(1, pos=(1, 1))
        g.add_edge(0, 1, distance=1.0, edge_id=0)
        with pytest.raises(ConfigurationError) as exc_info:
            # Provide edge_order so we get to the edge_spacing check
            Environment(track_graph=g, edge_order=[(0, 1)], edge_spacing=None)
        assert "edge_spacing" in str(exc_info.value)

    def test_valid_track_graph_minimal(self):
        """Test that valid track_graph configuration works."""
        g = nx.Graph()
        g.add_node(0, pos=(0, 0))
        g.add_node(1, pos=(1, 1))
        g.add_edge(0, 1, distance=1.0, edge_id=0)
        env = Environment(track_graph=g, edge_order=[(0, 1)], edge_spacing=0.0)
        assert env.track_graph is g


class TestEdgeOrderValidation:
    """Test validation of edge_order parameter."""

    def test_edge_order_must_be_list(self):
        """Test that edge_order must be a list."""
        g = nx.Graph()
        g.add_node(0, pos=(0, 0))
        g.add_node(1, pos=(1, 1))
        g.add_edge(0, 1, distance=1.0, edge_id=0)
        with pytest.raises(ValidationError) as exc_info:
            Environment(track_graph=g, edge_order=(0, 1), edge_spacing=0.0)
        assert "edge_order" in str(exc_info.value)
        assert "list" in str(exc_info.value).lower()

    def test_edge_order_elements_must_be_tuples(self):
        """Test that edge_order elements must be 2-tuples."""
        g = nx.Graph()
        g.add_node(0, pos=(0, 0))
        g.add_node(1, pos=(1, 1))
        g.add_edge(0, 1, distance=1.0, edge_id=0)
        with pytest.raises(ValidationError) as exc_info:
            Environment(track_graph=g, edge_order=[[0, 1]], edge_spacing=0.0)
        assert "edge_order[0]" in str(exc_info.value)
        assert "2-tuple" in str(exc_info.value)

    def test_edge_order_must_reference_existing_edges(self):
        """Test that edge_order must reference edges that exist."""
        g = nx.Graph()
        g.add_node(0, pos=(0, 0))
        g.add_node(1, pos=(1, 1))
        g.add_edge(0, 1, distance=1.0, edge_id=0)
        with pytest.raises(ValidationError) as exc_info:
            Environment(track_graph=g, edge_order=[(0, 2)], edge_spacing=0.0)
        assert "edge_order[0]" in str(exc_info.value)
        assert "non-existent" in str(exc_info.value).lower()

    def test_edge_order_accepts_reversed_edges(self):
        """Test that edge_order accepts edges in reverse order (undirected)."""
        g = nx.Graph()
        g.add_node(0, pos=(0, 0))
        g.add_node(1, pos=(1, 1))
        g.add_edge(0, 1, distance=1.0, edge_id=0)
        # Should work with reversed edge (1, 0) since graph is undirected
        env = Environment(track_graph=g, edge_order=[(1, 0)], edge_spacing=0.0)
        assert env.edge_order == [(1, 0)]


class TestEdgeSpacingValidation:
    """Test validation of edge_spacing parameter."""

    def test_edge_spacing_float_positive(self):
        """Test that edge_spacing as float works."""
        g = nx.Graph()
        g.add_node(0, pos=(0, 0))
        g.add_node(1, pos=(1, 1))
        g.add_edge(0, 1, distance=1.0, edge_id=0)
        env = Environment(track_graph=g, edge_order=[(0, 1)], edge_spacing=15.0)
        assert env.edge_spacing == 15.0

    def test_edge_spacing_negative_raises(self):
        """Test that negative edge_spacing raises."""
        g = nx.Graph()
        g.add_node(0, pos=(0, 0))
        g.add_node(1, pos=(1, 1))
        g.add_edge(0, 1, distance=1.0, edge_id=0)
        with pytest.raises(ValidationError) as exc_info:
            Environment(track_graph=g, edge_order=[(0, 1)], edge_spacing=-1.0)
        assert "edge_spacing" in str(exc_info.value)

    def test_edge_spacing_list_wrong_length_raises(self):
        """Test that edge_spacing list must have correct length."""
        g = nx.Graph()
        g.add_node(0, pos=(0, 0))
        g.add_node(1, pos=(1, 1))
        g.add_node(2, pos=(2, 2))
        g.add_edge(0, 1, distance=1.0, edge_id=0)
        g.add_edge(1, 2, distance=1.0, edge_id=1)
        with pytest.raises(ValidationError) as exc_info:
            Environment(
                track_graph=g,
                edge_order=[(0, 1), (1, 2)],
                edge_spacing=[15.0, 15.0],  # Should be length 1 (n_edges - 1)
            )
        assert "edge_spacing" in str(exc_info.value)
        assert "length" in str(exc_info.value).lower()

    def test_edge_spacing_list_correct_length(self):
        """Test that edge_spacing list with correct length works."""
        g = nx.Graph()
        g.add_node(0, pos=(0, 0))
        g.add_node(1, pos=(1, 1))
        g.add_node(2, pos=(2, 2))
        g.add_edge(0, 1, distance=1.0, edge_id=0)
        g.add_edge(1, 2, distance=1.0, edge_id=1)
        env = Environment(
            track_graph=g, edge_order=[(0, 1), (1, 2)], edge_spacing=[15.0]
        )
        assert env.edge_spacing == [15.0]

    def test_edge_spacing_list_element_negative_raises(self):
        """Test that negative values in edge_spacing list raise."""
        g = nx.Graph()
        g.add_node(0, pos=(0, 0))
        g.add_node(1, pos=(1, 1))
        g.add_node(2, pos=(2, 2))
        g.add_edge(0, 1, distance=1.0, edge_id=0)
        g.add_edge(1, 2, distance=1.0, edge_id=1)
        with pytest.raises(ValidationError) as exc_info:
            Environment(track_graph=g, edge_order=[(0, 1), (1, 2)], edge_spacing=[-1.0])
        assert "edge_spacing[0]" in str(exc_info.value)


class TestPositionRangeValidation:
    """Test validation of position_range parameter."""

    def test_position_range_must_be_sequence(self):
        """Test that position_range must be a sequence."""
        # String is a sequence, so it will fail on the element level
        # Let's test with a scalar instead
        with pytest.raises(ValidationError) as exc_info:
            Environment(position_range=123)
        assert "position_range" in str(exc_info.value)

    def test_position_range_elements_must_be_tuples(self):
        """Test that position_range elements must be 2-tuples."""
        with pytest.raises(ValidationError) as exc_info:
            Environment(position_range=[[0, 100]])
        assert "position_range[0]" in str(exc_info.value)
        assert "2-tuple" in str(exc_info.value)

    def test_position_range_must_contain_numeric_values(self):
        """Test that position_range must contain numeric values."""
        with pytest.raises(ValidationError) as exc_info:
            Environment(position_range=[("0", "100")])
        assert "position_range[0]" in str(exc_info.value)
        assert "numeric" in str(exc_info.value).lower()

    def test_position_range_min_must_be_less_than_max(self):
        """Test that min must be less than max in position_range."""
        with pytest.raises(ValidationError) as exc_info:
            Environment(position_range=[(100, 0)])
        assert "position_range[0]" in str(exc_info.value)
        assert "min" in str(exc_info.value).lower()
        assert "max" in str(exc_info.value).lower()

    def test_position_range_equal_min_max_raises(self):
        """Test that equal min and max raises."""
        with pytest.raises(ValidationError) as exc_info:
            Environment(position_range=[(0, 0)])
        assert "position_range[0]" in str(exc_info.value)

    def test_valid_position_range(self):
        """Test that valid position_range works."""
        env = Environment(position_range=[(0, 100), (0, 100)])
        assert env.position_range == [(0, 100), (0, 100)]


class TestFitPlaceGridValidation:
    """Test validation in fit_place_grid method."""

    def test_position_must_be_ndarray(self):
        """Test that position must be numpy array."""
        env = Environment()
        with pytest.raises(ValidationError) as exc_info:
            env.fit_place_grid(position=[[0, 0], [1, 1]])
        assert "position" in str(exc_info.value)
        assert "numpy" in str(exc_info.value).lower()

    def test_position_must_be_finite(self):
        """Test that position must contain finite values."""
        from non_local_detector.exceptions import DataError

        env = Environment()
        position = np.array([[0, 0], [np.nan, 1], [2, 2]])
        with pytest.raises(DataError) as exc_info:
            env.fit_place_grid(position=position)
        assert "position" in str(exc_info.value)
        assert "nan" in str(exc_info.value).lower()

    def test_position_must_be_1d_or_2d(self):
        """Test that position must be 1D or 2D."""
        env = Environment()
        position = np.array([[[0, 0], [1, 1]]])  # 3D
        with pytest.raises(ValidationError) as exc_info:
            env.fit_place_grid(position=position)
        assert "position" in str(exc_info.value)
        assert "1D or 2D" in str(exc_info.value)

    def test_position_cannot_be_empty(self):
        """Test that position cannot be empty."""
        env = Environment()
        position = np.array([]).reshape(0, 2)
        with pytest.raises(ValidationError) as exc_info:
            env.fit_place_grid(position=position)
        assert "position" in str(exc_info.value)
        assert "no time points" in str(exc_info.value).lower()

    def test_valid_position_1d(self):
        """Test that valid 1D position works."""
        env = Environment()
        # Need to reshape to 2D for fit_place_grid or provide 2D from the start
        position = np.array([[0], [1], [2], [3], [4]])
        result = env.fit_place_grid(position=position)
        assert result is env
        assert env.place_bin_centers_ is not None

    def test_valid_position_2d(self):
        """Test that valid 2D position works."""
        env = Environment()
        position = np.array([[0, 0], [1, 1], [2, 2], [3, 3]])
        result = env.fit_place_grid(position=position)
        assert result is env
        assert env.place_bin_centers_ is not None


class TestIsTrackInteriorValidation:
    """Test validation of is_track_interior parameter."""

    def test_is_track_interior_must_be_ndarray(self):
        """Test that is_track_interior must be numpy array."""
        with pytest.raises(ValidationError) as exc_info:
            Environment(is_track_interior=[[True, False], [False, True]])
        assert "is_track_interior" in str(exc_info.value)
        assert "numpy" in str(exc_info.value).lower()

    def test_valid_is_track_interior(self):
        """Test that valid is_track_interior works."""
        is_interior = np.array([True, False, True, False])
        env = Environment(is_track_interior=is_interior)
        assert np.array_equal(env.is_track_interior, is_interior)
