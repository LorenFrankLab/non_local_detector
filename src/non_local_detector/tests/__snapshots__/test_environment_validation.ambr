# serializer version: 1
# name: test_environment_1d_snapshot
  dict({
    'edges': list([
      dict({
        'dim': 0,
        'max': 10.0,
        'min': 0.0,
        'n_edges': 11,
        'values': list([
          0.0,
          1.0,
          2.0,
          3.0,
          4.0,
          5.0,
          6.0,
          7.0,
          8.0,
          9.0,
          10.0,
        ]),
      }),
    ]),
    'environment_name': 'test_1d',
    'is_track_interior': dict({
      'dtype': 'bool',
      'n_false': 0,
      'n_true': 10,
      'shape': tuple(
        10,
      ),
      'values': list([
        True,
        True,
        True,
        True,
        True,
        True,
        True,
        True,
        True,
        True,
      ]),
    }),
    'place_bin_centers': dict({
      'dtype': 'float64',
      'first_5': list([
        list([
          0.5,
        ]),
        list([
          1.5,
        ]),
        list([
          2.5,
        ]),
        list([
          3.5,
        ]),
        list([
          4.5,
        ]),
      ]),
      'last_5': list([
        list([
          5.5,
        ]),
        list([
          6.5,
        ]),
        list([
          7.5,
        ]),
        list([
          8.5,
        ]),
        list([
          9.5,
        ]),
      ]),
      'max': 9.5,
      'mean': 5.0,
      'min': 0.5,
      'shape': tuple(
        10,
        1,
      ),
      'std': 2.8722813232690143,
    }),
    'place_bin_edges': dict({
      'dtype': 'float64',
      'first_5': list([
        list([
          0.0,
        ]),
        list([
          1.0,
        ]),
        list([
          2.0,
        ]),
        list([
          3.0,
        ]),
        list([
          4.0,
        ]),
      ]),
      'max': 10.0,
      'mean': 5.0,
      'min': 0.0,
      'shape': tuple(
        11,
        1,
      ),
    }),
    'place_bin_size': 1.0,
  })
# ---
# name: test_environment_2d_snapshot
  dict({
    'edges': list([
      dict({
        'dim': 0,
        'max': 22.0,
        'min': -2.0,
        'n_edges': 13,
        'values': list([
          -2.0,
          0.0,
          2.0,
          4.0,
          6.0,
          8.0,
          10.0,
          12.0,
          14.0,
          16.0,
          18.0,
          20.0,
          22.0,
        ]),
      }),
      dict({
        'dim': 1,
        'max': 16.875,
        'min': -1.875,
        'n_edges': 11,
        'values': list([
          -1.875,
          0.0,
          1.875,
          3.75,
          5.625,
          7.5,
          9.375,
          11.25,
          13.125,
          15.0,
          16.875,
        ]),
      }),
    ]),
    'environment_name': 'test_2d',
    'is_track_interior': dict({
      'dtype': 'bool',
      'n_false': 40,
      'n_true': 80,
      'shape': tuple(
        12,
        10,
      ),
      'values': 'too_large',
    }),
    'place_bin_centers': dict({
      'dtype': 'float64',
      'first_5': list([
        list([
          -1.0,
          -0.9375,
        ]),
        list([
          -1.0,
          0.9375,
        ]),
        list([
          -1.0,
          2.8125,
        ]),
        list([
          -1.0,
          4.6875,
        ]),
        list([
          -1.0,
          6.5625,
        ]),
      ]),
      'last_5': list([
        list([
          21.0,
          8.4375,
        ]),
        list([
          21.0,
          10.3125,
        ]),
        list([
          21.0,
          12.1875,
        ]),
        list([
          21.0,
          14.0625,
        ]),
        list([
          21.0,
          15.9375,
        ]),
      ]),
      'max': 21.0,
      'mean': 8.75,
      'min': -1.0,
      'shape': tuple(
        120,
        2,
      ),
      'std': 6.3164694615214705,
    }),
    'place_bin_edges': dict({
      'dtype': 'float64',
      'first_5': list([
        list([
          -2.0,
          -1.875,
        ]),
        list([
          -2.0,
          0.0,
        ]),
        list([
          -2.0,
          1.875,
        ]),
        list([
          -2.0,
          3.75,
        ]),
        list([
          -2.0,
          5.625,
        ]),
      ]),
      'max': 22.0,
      'mean': 8.75,
      'min': -2.0,
      'shape': tuple(
        143,
        2,
      ),
    }),
    'place_bin_size': tuple(
      2.0,
      2.0,
    ),
  })
# ---
# name: test_environment_edges_snapshot
  dict({
    'edges': list([
      dict({
        'dim': 0,
        'max': 25.0,
        'min': 0.0,
        'n_edges': 11,
        'values': list([
          0.0,
          2.5,
          5.0,
          7.5,
          10.0,
          12.5,
          15.0,
          17.5,
          20.0,
          22.5,
          25.0,
        ]),
      }),
    ]),
    'environment_name': 'test_edges',
    'is_track_interior': dict({
      'dtype': 'bool',
      'n_false': 0,
      'n_true': 10,
      'shape': tuple(
        10,
      ),
      'values': list([
        True,
        True,
        True,
        True,
        True,
        True,
        True,
        True,
        True,
        True,
      ]),
    }),
    'place_bin_centers': dict({
      'dtype': 'float64',
      'first_5': list([
        list([
          1.25,
        ]),
        list([
          3.75,
        ]),
        list([
          6.25,
        ]),
        list([
          8.75,
        ]),
        list([
          11.25,
        ]),
      ]),
      'last_5': list([
        list([
          13.75,
        ]),
        list([
          16.25,
        ]),
        list([
          18.75,
        ]),
        list([
          21.25,
        ]),
        list([
          23.75,
        ]),
      ]),
      'max': 23.75,
      'mean': 12.5,
      'min': 1.25,
      'shape': tuple(
        10,
        1,
      ),
      'std': 7.180703308172536,
    }),
    'place_bin_edges': dict({
      'dtype': 'float64',
      'first_5': list([
        list([
          0.0,
        ]),
        list([
          2.5,
        ]),
        list([
          5.0,
        ]),
        list([
          7.5,
        ]),
        list([
          10.0,
        ]),
      ]),
      'max': 25.0,
      'mean': 12.5,
      'min': 0.0,
      'shape': tuple(
        11,
        1,
      ),
    }),
    'place_bin_size': 2.5,
  })
# ---
# name: test_environment_with_track_interior_mask_snapshot
  dict({
    'edges': list([
      dict({
        'dim': 0,
        'max': 12.0,
        'min': 0.0,
        'n_edges': 9,
        'values': list([
          0.0,
          1.5,
          3.0,
          4.5,
          6.0,
          7.5,
          9.0,
          10.5,
          12.0,
        ]),
      }),
    ]),
    'environment_name': 'test_masked',
    'is_track_interior': dict({
      'dtype': 'bool',
      'n_false': 0,
      'n_true': 8,
      'shape': tuple(
        8,
      ),
      'values': list([
        True,
        True,
        True,
        True,
        True,
        True,
        True,
        True,
      ]),
    }),
    'place_bin_centers': dict({
      'dtype': 'float64',
      'first_5': list([
        list([
          0.75,
        ]),
        list([
          2.25,
        ]),
        list([
          3.75,
        ]),
        list([
          5.25,
        ]),
        list([
          6.75,
        ]),
      ]),
      'last_5': list([
        list([
          5.25,
        ]),
        list([
          6.75,
        ]),
        list([
          8.25,
        ]),
        list([
          9.75,
        ]),
        list([
          11.25,
        ]),
      ]),
      'max': 11.25,
      'mean': 6.0,
      'min': 0.75,
      'shape': tuple(
        8,
        1,
      ),
      'std': 3.43693177121688,
    }),
    'place_bin_edges': dict({
      'dtype': 'float64',
      'first_5': list([
        list([
          0.0,
        ]),
        list([
          1.5,
        ]),
        list([
          3.0,
        ]),
        list([
          4.5,
        ]),
        list([
          6.0,
        ]),
      ]),
      'max': 12.0,
      'mean': 6.0,
      'min': 0.0,
      'shape': tuple(
        9,
        1,
      ),
    }),
    'place_bin_size': 1.5,
  })
# ---
